{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "WBT-MBT2-25E <i>Programming Python for Bioinformatics</i> &copy; 2021 Adrian Kania (adrian15x.kania@uj.edu.pl) Department of Computational Biophysics and Bioinformatics, Faculty of Biochemistry, Biophysics and Biotechnology, Jagiellonian University\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddffee;border-radius:15px\">Mini project: Sequences comparison using an alignment and free-alignment methods </p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will consider neuraminidase (N) sequences from the avian influenza viruses. This project aims to compare the effectiveness of alignment and free-alignment approaches for sequences comparison. The letter N and H refers to neuraminidase and hemagglutinin type respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddffee;border-radius:15px\">1. Neuraminidase sequences </p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download sequences and IDs from $influenza\\_viruses$ FASTA file and save them into two lists. Convert nucleotides to upper case. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "IDS = []\n",
    "sequences = []\n",
    "start_pattern = '>.*'\n",
    "\n",
    "with open('influenza_viruses', 'r') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        if re.match(start_pattern, line):\n",
    "            IDS.append(line)\n",
    "        else:\n",
    "             sequences.append(line.upper())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To perform steps 2,3 and 4 I recommend creating three separate scripts named: step2.py, step3.py and step4.py in which you append all necessary functions and variables.  <br>\n",
    "\n",
    "- step2.py should contain: DNA dictionary (points for matches and mismatches), SequenceAlign, Profile, ProfileAlign, and the main functions: ProfileMultipleAlignment, Score, SimMatrix\n",
    "- step3.py should contain: MerDict, CompareSequences, MerMatrix\n",
    "- step4.py should contain: WordSeq, LZcomplexity, LZMatrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddffee;border-radius:15px\">2. Multiple Sequence Alignment (MSA)  </p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['DNA_2', 'Profile', 'ProfileAlign', 'ProfileMultipleAlignment', 'Score', 'SequenceAlign', 'SimMatrix', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__']\n"
     ]
    }
   ],
   "source": [
    "import step2\n",
    "print(dir(step2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from step2 import DNA_2, SequenceAlign, Profile, ProfileAlign, ProfileMultipleAlignment, Score, SimMatrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use ProfileMultipleAlignment function to perform the MSA. Put the following points:\n",
    "- match (2), mismatch (-2) (DNA_2 dictionary)\n",
    "- gap insertion/gap extension (4)\n",
    "\n",
    "Save it to an $align$ variable. \n",
    "\n",
    "Use SimMatrix to compare aligned sequences.\n",
    "\n",
    "Build a UPGMA tree and plot it. Make a comment on it. Whether viruses are classified appropriately?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddffee;border-radius:15px\">3. K-mers </p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CompareSequences', 'DNA_2', 'MerDict', 'MerMatrix', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__']\n"
     ]
    }
   ],
   "source": [
    "import step3\n",
    "print(dir(step3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from step3 import MerDict, CompareSequences, MerMatrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use MerMatrix function (with k=3) to compare viruses sequences and use this matrix to construct a phylogenetic tree (UPGMA method). Plot a tree and make a comment on it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddffee;border-radius:15px\">4. Lempel-Ziv Complexity </p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['DNA_2', 'LZMatrix', 'LZcomplexity', 'WordSeq', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__']\n"
     ]
    }
   ],
   "source": [
    "import step4\n",
    "print(dir(step4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from step4 import WordSeq, LZcomplexity, LZMatrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use LZMatrix function to compare viruses sequences and use this matrix to construct a phylogenetic tree (UPGMA method). Plot a tree and make a comment on it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddffee;border-radius:15px\">5. Time consumption </p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run code from steps 2, 3 and 4 once again but this time measure the time needed to perform calculations. Which method is the fastest, which the slowest?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7230541706085205\n"
     ]
    }
   ],
   "source": [
    "#To measure the execution time you may use time library and calculate the difference between the time before and after part of your code. Below an example:\n",
    "\n",
    "import time\n",
    "start_time = time.time()\n",
    "\n",
    "#############Your code here##########\n",
    "x = [1,2]\n",
    "N = 150000\n",
    "for i in range(N):\n",
    "    x.append(x[-1]+x[-2])\n",
    "#####################################\n",
    "\n",
    "end_time = time.time()\n",
    "seconds = end_time - start_time\n",
    "print(seconds)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddffee;border-radius:15px\">6. H5N1 viruses and 20 aa deletion </p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It appears that many H5N1 subtype viruses posses a characterisitc 60 nt (20 aa) deletion which causes\n",
    "higher virulence of these viruses. Download neuraminidase protein sequences for A/mandarin_duck/Korea/K10-483/2010(H5N1) [ID: JF699677.1] and A/duck/Hokkaido/w73/2007(H1N1) [ID: AB470663.1] from Nucleotide database finding an appropriate tag (via Python/eutils). Plot a dot matrix and localise this deletion. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "ID = ['AEA02263', 'BAG85128']\n",
    "seqs = []\n",
    "for elem in ID:\n",
    "    handle = Entrez.efetch(db=\"protein\", id=elem, rettype=\"gb\", retmode=\"text\")\n",
    "    record = SeqIO.read(handle, \"genbank\")\n",
    "    handle.close()\n",
    "    seqs.append(record.seq)\n",
    "print(seqs)\n",
    "print(seqs[1])\n",
    "print(seqs[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"font-size:15pt;font-weight:bold;border:1px solid;border-color:#aabbcc;padding:15px;background:#ddffee;border-radius:15px\">7. Neuraminidase inhibitors </p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are many drugs classified as neuraminidase inhibitors (NAIs) and one of them is zanamivir. Download the zanamivir SMILES structure from the Pubchem database and plot it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from urllib.request import urlopen\n",
    "from lxml import etree\n",
    "\n",
    "molecule = \"zanamivir\"\n",
    "query = f\"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/name/{molecule}/XML\"\n",
    "\n",
    "response = urlopen(query)\n",
    "resultxml = response.read()\n",
    "tree = etree.XML(resultxml)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "nam = []\n",
    "for element in tree.iter():\n",
    "    if element.tag == \"{http://www.ncbi.nlm.nih.gov}PC-InfoData_value_sval\": \n",
    "        nam.append(element.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['(2R,3R,4S)-3-acetamido-4-guanidino-2-[(1R,2R)-1,2,3-trihydroxypropyl]-3,4-dihydro-2H-pyran-6-carboxylic acid', 'InChI=1S/C12H20N4O7/c1-4(18)15-8-5(16-12(13)14)2-7(11(21)22)23-10(8)9(20)6(19)3-17/h2,5-6,8-10,17,19-20H,3H2,1H3,(H,15,18)(H,21,22)(H4,13,14,16)/t5-,6+,8+,9+,10+/m0/s1', 'ARAIBEBZBOPLMB-UFGQHTETSA-N', '332.13319899', 'C12H20N4O7', '332.31', 'CC(=O)NC1C(C=C(OC1C(C(CO)O)O)C(=O)O)N=C(N)N', 'CC(=O)N[C@@H]1[C@H](C=C(O[C@H]1[C@@H]([C@@H](CO)O)O)C(=O)O)N=C(N)N', '332.13319899']\n"
     ]
    }
   ],
   "source": [
    "print(nam[5:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAMAAAE5pE7RAAABd1BMVEX19fVmi4svT08AAP//AADv8PC9zMx+nZ2wwsKptbV7j49ddXXHz8/X3NxNaGiZqKg+W1vm6em4wsJsgoKioviLnJzv7/VJSfz+CwuWlvj3tbX17e34oKD+AQH6bW313d3+AwP5jo74o6PZ2fazvr5xhob17+/4lJT5kpL3sbH3r6/9GRn+ExP3s7P+FRX7WFj6dXX7UVH8Ozv8RUX15eX8SUn8Pz/21NT229v+CQn+Bwf8Njb6aGj6YmL8MjL9Kir5e3v4lpb4mpr5fX39MDD5hYX5gYH22dn6bGz22NiTk/liYvsxMf3ExPcJCf4bG/1YWPsBAf5UVPvCwvc3N/wHB/53d/rr6/UeHv0ZGf2YmPiOjvkmJv0wMP38UlL7YmL5k5P3xMT9MTH6e3v6enqElpagrq7Z3t5LZmZof3+8xcV7e/p6evppafuMjPnS0vYjI/5GRvyvr/idq6uHmZlFYWHf4+Nac3PK0dH8PT34uLj3uLj9PT1oUWjPAAAACXBIWXMAAA7EAAAOxAGVKw4bAAAKmUlEQVR4nO2diZ/TthLHlTj7INlsjg3JLteDchV60EJLW3rT673HUXpAKaXcFEqh93388c93ZFsje7S2oyS/7+eza8uS7PGsV5ZmpLEQYKG5MZASK/9SFzra8jeX5cTsuer/nuzwN+PgWNv/7d9TIy7YE6vNuFA3Ovog2EyLadizFTn1fCwnjmsKbit2vrhYV1dqPhjcKFjw3yVdcHtJ5wGgPqhmWyZsJrMUav9KxQk2V4JN0GKHzXf8MpKkitsxufU+LZ2vTV5pn7GMepzsoUzL7d2B35wMM4dVpYQY0aUWg7IaaY+CL0MAAAhwXx7r4e6KkN6ZiUQ+L78vboe7ZydiRzKHyaa0nzzXZrqoglJfEyWeLPF+nFyiuh65pET6IZHqeb+GqkQn/qW+q678Tu5lzjlNPJgmWqW+vuaMxN+zLQ7F+yfkjERCRaOR3Po0RTOnOKiO9VLsBHtfj3dvywkTdr1JJWbLNtgGAAAAgKWE6qnyGYnVTrSf6FmYdjMGWz6X12+7GSWcnakcHomx0sY1KqcQiT7gxnUqh6TMUQRGJCwceeB4cSdZrtjJpP0NOdEeUwmZRnLnvJx3Ltz6ZoGeKuEbjiLjUfpcAWvRji9Bl0w8DBN/ipMHhOqRGvcTvgDaQPFQTpzJnMfHpMk6RXj4y2v+7D2XwrVjA7RYR8R+IkGeK3WyX+K9Y0L6yycSKqInVzafjnR/DE3rGdl6VhpZU6zqjDrbkEou4f8TjhSl9XIBUJgS26HOKvWwchnkFynE1Ldx08TRIXNUblmO5jUzOWySiYWB6Z+q7VwAAAAAAAAAAACYA3pD0c8cXNkmtqttBHROeYzbQqz7Qq2uJTJmK5boyNJIzqOZiLXrOW9CqspwNrl4aad6um1Yp0p2bwrC/zDZuOKoxaLrMGnYacdvNKycxmqptmyVyk6xnI8ov+DEEUQzUAPOwc8psTYuX6tdrOhRd9Re1J++dcUS15VidX0fzEgxQ5rOCd5ba6uEFNMDqYbhrLSMcSBkWS8k/EomifvyW77/t1as7LM+9O61s+6+mzO3OxrGBZg5amfZ07vFW/9RZRA0kxMiJNrkqlBNzn0iw31vvvsESywrgVgc5m4myAxxDhrI9ezma+wcHs4nO9hiPfk81U+mc5g44guNWF6L9vtvmcN7X3njA3UFOqcY6SafHLh47f+h6eIvcdfo9KWL5TJ0lRYI1hsSRcoSi0t7PKYDFAAAgI30/VbVKoJRVX9dM827duQewVqHLFYXlJmvVYcBkIY285VmAARVU4uhnQ/EAgAAAAAAAAAAAAAAAACArRPOIBYHiMUBYnGAWBwgFgeIxcFOsfo2irXeUWmLnuZaywRYb8bWWic7kWy2Yg3Cn3AZT5GLVy+W963DfhRiNDm/bcZ/RHkt07kLRS5eoVhPvbDPn8V2J3Xc0SyHi+pUiHv+t58Q4lY2x3EoscI6VeJe4p3/KnPoVXp0ndJ45kXx3v+UOY64SohF1ymPl14lvvRKLB7U1uFh6XQxQ7Ho9S3l/COuNBqKSLG5VC6WpdqyUyxD6M7XTLtlEIvDkoqlWZhPd19qEIsMG6LpvtQgFh0vgO6+1CEWFc1E031xBxxUKGY6R7V4MNOYRweoMAY/aLov6mAF+hx17ANSrBB54ODy7U/xrpPMCWIN+Kyn19PQOdMhZnIFaFasdM9Beia8Udejx1ORZd23EzEUiuXQIRnytOXSDYeAnrrbjxNZPcVe8ZyHD5I590S84rqAWEEAB1/dmVF9GO5hlF2Dm5tzXxV74chx1tqgXlN9CX/lcle9Ykyb8+ChMoe7ELxNXcKsqaJyjp1gdszKvLgmZ6vr07d08Tq6ExCLA8TiMHdi0S1QDYNGiMVhdiOtyYea+GSkWCdPH2DnsJg4n/LFomNWcaNZkWJtOHyx3OfnjPpje3QOV6zPvtSK9av64kQoLW6QLVosQY3ORNCr/uNw5vCeU5SHgM4pRqYHTfhGvH7rz4cUGczTlyxW0MuXH7J/KhUrPTqj42d6w644GlH/L7PTFxWrsLZEMEj1FZYZpBY+fRVi6Yb05YrFhjaAzBYb3fwCYvGAWBwgFgeIxQFicYBYHGwVK/t9BRtoTj0p9tAbeh9ZSH32YdYEbq54wqAdxE5B/0sZN2YpypS796RE8UFGZagn/d0hc2qCmPR3q5bpgDT0pL8apgPS0JP+6pgOSENP+itpOuBiMdP5YjSWimUlln4xBmJxgFgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACwEFi6iNJOoCwGUBYDKIsBlMUAymIAZTGAshhAWQygLAZQFgMoiwGUxQDKYgBlFaQ9GBdT1kpjm/t7OyvUqUkdW+mujvwoUu3mIPP5lAzLrKxhpxfEARv3/DhXa6PVrraCF0C3wQyia1LHOg6EkdyGo8QHRtZ7HdX3RgKW9MlqBbHYuz3VR0/6zbPqWkukrGdPbp5+TQpYf6tJt1IXzp8TXmD6Dw8KsVMT17zIdeaQI08f3+1u3tofxNG/eed2fp2J43z8KVNZqevMJ0+2nvc2Z1qM6P4TZ+MTZwdPWSbXsY5jrRPe5t0WI+SeqyzxmfMlS1km17GPPXtPvbLvjRbnO06essQXDq/NMrgOsIR638b0x2lpLIr3B2UxgLIYQFkMoCxrMTEL2mZKrC2qJZTFYBGUVRtzriwjy4mZuUUsgLL4lhOzSh5zryy+5cSsksf8K4ttOTGrJDwbM/+rKyZ1ClZi9zHZlpPvHDNzS+i9iL1AVdVhVMpVlraAQzgXIr7/8cfvcwRQkvKLRf7F0uswK+V61nK0GTgXlPSb3/nbR+kxy9nz53VKHis9rp7nutw6/Epbe7ICOtmvE0UewEfO4+KVhPcdTt3XQcurI775+itNpX/uKg+XoSyX/mD6/x492vqbSFXyWB/kt7Jl1BFfff1NXp2//OH4U3tfeHnf6633cy8RUvgNEIgdPdpFbiJRLHNHFdXRP1JJdrWe8zZvV+L/iJ6l4jcRViL+V0qvU+CRkjkauNTeaVX0waUm+yZEfV1Pbp3dz2y+6G7e21/RZ6CMus+2Ksv9R3xp89U3K/s21YIpq1qgLAZQFgMoiwGUxQDKYmDi6DO6CRPvoG0eRSiLAZRVNba1JTUwcZyPjPxdbGVxLSemdSpk4lx0LtWhLBPLSaXWFgMmzsbBy59fM1ZW8+eCxU0sJ1VbW7h4LpwrV5kzZgNCs85vh3//o0BpE8tJ1dYWLr6/S1w3UZaYGgwP/fpLfmETy0m11pY89E4yh++TjG2/hw9TRf4u7e1f87KeknwdCWInh1Jff95jCahlAZTlEfk8kvrK8QGyWRBlCUPvMo/FUZaH0byF4szpUkQNRjNilpYlHDWaA2UxgLIYQFkMoCwGUBYDKIsBlMUAymIAZTGAshhAWQygLAZQFoMmsQwCpGkPgiBbuRHulp3IkhyFb9NGuFtmUpbkODBgvznfIdfK567SkhyGnBQHFmLWkDlylL6WxjHYbt5S1lkiWFH6bt+5ya6zSJhE6VuIyH4mmETpW4zIfiaYROlDZD9QjCXvD/CAskDpLN5klgqBshgsxLcA6gJPFgMoCwAAAAAAAAAAAAAAAMBS8X+v0cS97MCgVwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from urllib.request import urlretrieve\n",
    "from IPython.display import Image\n",
    "\n",
    "\n",
    "smiles = nam[-3]\n",
    "urlretrieve('https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles/'+smiles+'/PNG', 'smi_pic.png')\n",
    "plot = Image(filename='smi_pic.png')\n",
    "plot"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
